test/rulesets/Base/mapreduce.jl

```diff
@@ -20,4 +20,73 @@
     end  # sum abs2
+
+    @testset "prod" begin
+        @testset "Array{$T}" for T in [Float64, ComplexF64]
+            @testset "size = $sz, dims = $dims" for (sz, dims) in [
+                ((12,), :), ((12,), 1),
+                ((3,4), 1), ((3,4), 2), ((3,4), :), ((3,4), [1,2]),
+                ((3,4,1), 1), ((3,2,2), 3), ((3,2,2), 2:3),
+                ]
+                x = randn(T, sz)
+                test_rrule(prod, x; fkwargs=(dims=dims,), check_inferred=true)
+                x[1] = 0
+                test_rrule(prod, x; fkwargs=(dims=dims,), check_inferred=true)
+                x[5] = 0
+                test_rrule(prod, x; fkwargs=(dims=dims,), check_inferred=true)
+                x[3] = x[7] = 0  # two zeros along some slice, for any dims
+                test_rrule(prod, x; fkwargs=(dims=dims,), check_inferred=true)
+
+                if ndims(x) == 3
+                    xp = PermutedDimsArray(x, (3,2,1))  # not a StridedArray
+                    xpdot, xpbar = permutedims(rand(T, sz), (3,2,1)), permutedims(rand(T, sz), (3,2,1))
+                    test_rrule(prod, xp ⊢ xpbar; fkwargs=(dims=dims,), check_inferred=true)
+                end
+            end
+
+            @testset "structured wrappers" begin
+                # Adjoint -- like PermutedDimsArray this may actually be used
+                xa = adjoint(rand(T,4,4))
+                test_rrule(prod, xa ⊢ rand(T,4,4))
+                @test_skip test_rrule(prod, xa ⊢ rand(T,4,4), fkwargs=(dims=2,)) # seems broken?
+#=
+xa = randn(3,3)'
+Zygote.gradient(x -> sum(prod(x,dims=1)), xa)[1]
+ForwardDiff.gradient(x -> sum(prod(x,dims=1)), xa)
+Zygote.gradient(x -> sum(prod(x,dims=1)), Matrix(xa))[1]
+# These all agree, so the fault is in the testing, somehow.
+=#
+                @test unthunk(rrule(prod, adjoint(rand(T,3,3)))[2](1.0)[2]) isa Matrix
+                @test unthunk(rrule(prod, adjoint(rand(T,3,3)), dims=1)[2](ones(1,3))[2]) isa Matrix
+                # Diagonal -- a stupid thing to do, product of zeros! Shouldn't be an error though:
+                @test iszero(unthunk(rrule(prod, Diagonal(rand(T,3)))[2](1.0)[2]))
+                @test iszero(unthunk(rrule(prod, Diagonal(rand(T,3)), dims=1)[2](ones(1,3))[2]))
+                @test unthunk(rrule(prod, Diagonal(rand(T,1)))[2](1.0)[2]) == hcat(1) # 1x1 sparse matrix
+                @test unthunk(rrule(prod, Diagonal(ones(T,2)), dims=1)[2](ones(1,2))[2]) == [0 1; 1 0]
+                # Triangular -- almost equally stupud
+                @test iszero(unthunk(rrule(prod, UpperTriangular(rand(T,3,3)))[2](1.0)[2]))
+                @test unthunk(rrule(prod, UpperTriangular(ones(T,2,2)))[2](1.0)[2]) == [0 0; 1 0]
+                # Symmetric -- at least this doesn't have zeros, still an unlikely combination
+                xs = Symmetric(rand(T,4,4))
+                @test_skip test_rrule(prod, xs ⊢ rand(T,4,4))
+                @test_skip test_rrule(prod, xs ⊢ rand(T,4,4), fkwargs=(dims=2,))
```

Comment:
Is there a bug in how FiniteDifferences does this, or am I thinking incorrectly about what it should produce?
```
using ForwardDiff, FiniteDifferences
xs = Symmetric(reshape(1:16,4,4)./10)
xm = Matrix(xs)

g1 = ForwardDiff.gradient(prod, xm) # symmetric, but not Symmetric
g2 = grad(central_fdm(5, 1), prod, xm)[1]
g3 = grad(central_fdm(5, 1), prod, xs)[1]  # is Symmetric, and differs

g1 ≈ g2
diag(g1) ≈ diag(g3)
UnitUpperTriangular(g1) ≈ UnitUpperTriangular(g3 ./ 2)  # this seems weird
```
With `dims`:
```
g4 = ForwardDiff.gradient(x -> sum(prod(x,dims=1)), xm) # no longer symmetric
g5 = grad(central_fdm(5, 1), x -> sum(prod(x,dims=1)), xm)[1] 
g6 = grad(central_fdm(5, 1), x -> sum(prod(x,dims=1)), xs)[1] 

g4 ≈ g5

proj(m) = (m .+ m')./2;
proj(g4) ≈ proj(proj(g4)) # it's a projection

fold(m) = m .+ m' .- Diagonal(m)
fold(g4) ≈ g6
```

Reply:
@wesselb

Reply:
I suggest checking what `FiniteDifferences.to_vec` is output ting

Reply:
For now I've left here a simpler test that it does run without error on a Symmetric: `unthunk(rrule(prod, Symmetric(ones(T,2,2)))[2](1.0)[2]) == [1 1; 1 1]`. I very much doubt this case is going to see use, but it shouldn't give an error.

Reply:
Would it be better to have a test with `Symmetric([2.0 3.0; 3.0 2.0])` ones are hard to trust

Reply:
OK, I've switched it. It's mostly to check this doesn't give an error, the computation here does not care at all what kind of matrix it gets. Although `ones` is in fact sufficient to see this weirdness:
```
julia> grad(central_fdm(5, 1), prod, Symmetric(ones(2,2)))[1]
2×2 Symmetric{Float64, Matrix{Float64}}:
 1.0  2.0
 2.0  1.0

julia> ForwardDiff.gradient(prod, ones(2,2))
2×2 Matrix{Float64}:
 1.0  1.0
 1.0  1.0
 ```